### ISSUE TITLE
Default SeoMetadata class

### ISSUE THREAD

**Issue by dbu** [symfony-cmf/SeoBundle#201]
@goemaere found that when using the [conference tutorial](https://github.com/dbu/conference-tutorial/pull/16) and running in production mode, creating a new page and then loading it again fails. after some debugging, i found that this is because the sonata extension [form type sets data_class to the model class](https://github.com/symfony-cmf/SeoBundle/blob/master/Form/Type/SeoMetadataType.php#L64). this does not work out because we store that in the db and get a doctrine error that there is no proxy class available. (in terms of doctrine, the model class is not complete, having no idea - there is probably a missing doctrine sanity check, it should prevent storing this if it does not work)

either we need to document that the user MUST init SeoMetadata to the Doctrine\Phpcr class in the page constructor or we make the default class configurable and inject it to the form type. i prefer the later.

**Comment by kminh** [comment]
Not sure if related to this issue, but using version 1.1 of this bundle, when go into production this file is missing: `app/cache/prod/__CG__SymfonyCmfBundleSeoBundleModelSeoMetadata.php`.

Do you have any work around for now? Currently I have to copy this file from the dev environment.

**Comment by ElectricMaxxx** [comment]
Does look related. Looks like you forgot to clear cache for prod Environment.

**Comment by kminh** [comment]
I did, but that proxy class doesn't get generated when clearing the cache. When accessing a page with SeoMetadata attached in dev environment that proxy class is generated.

**Comment by dbu** [comment]
i think this is exactly the issue. can you check what class is stored in the class property of the node? `app/console doctrine:phpcr:node:dump /path/to/document --properties --depth=0` if it is the Model\SeoMetadata, that is the problem. 

you could run a phpcr-shell query to update the class name to the one in Doctrine\Phpcr.

and to prevent this from happening in the future, you need to create the metadata in the constructor of your documents as the form autocreates it to the model class which will not work out.

**Comment by kminh** [comment]
`app/console doctrine:phpcr:node:dump /cms/content/pages/name/seoMetadata --props --depth=0` returns: `phpcr:class = Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata`

By creating the metadata in the constructor, do you mean this:
```php
// in Acme/AcmeBundle/Doctrine/Phpcr/Page
use Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetada;

class Page
{
    public function __constructor()
    {
        $this->seoMetadata = new SeoMetada();
    }
}
```

I tried that and it didn't seem to work.

**Comment by dbu** [comment]
exactly. but this will only work for new pages. for the existing, you need something like 

```
PHPCRSH > UPDATE [nt:unstructured] SET phpcr:class="Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetada" WHERE phpcr:class="Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata"
```

see http://phpcr.readthedocs.org/en/latest/phpcr-shell/installation.html#install-as-an-embedded-application for how to install the shell. it is pretty new, but really helpful.

**Comment by kminh** [comment]
That works, thanks @dbu.

Would it cause any issue in the future to not have `phpcr:classparents` set to `Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata`? For new pages I see this:

```
phpcr:class => Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetadata
phpcr:classparents => Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata
```

**Comment by dbu** [comment]
glad it works! so we really should see how to help the form system to do
the right thing automatically.

the parents are only relevant when you search for documents of the
parent class, which you probably don't in this case. the parents are
also updated automatically if you ever edit the metadata document again.

**Comment by ConneXNL** [comment]
Just ran into this problem too while adding content in sonataadminbundle.

**Comment by liquorvicar** [comment]
@dbu We're still having issues with this. We've set the MetaData explicitly in the the BasePage but we're getting the same error whenever we clear the cache in production. We have a local vagrant VM set up and that seems fine, even when running `cache:clear --env=prod`. I'm guessing that there is some node in the database that still has the old class set. I've tried installing PHPCRSH on staging (which has a copy of our production db) but I'm hitting this error:
```
PHPCRSH > SELECT * FROM [nt:unstructured]  WHERE phpcr:class="Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata"
[Doctrine\DBAL\Exception\InvalidFieldNameException] An exception occurred while executing 'SELECT n0.id AS n0_id, n0.path AS n0_path, n0.parent AS n0_parent, n0.local_name AS n0_local_name, n0.namespace AS n0_namespace, n0.workspace_name AS n0_workspace_name, n0.identifier AS n0_identifier, n0.type AS n0_type, n0.props AS n0_props, n0.numerical_props AS n0_numerical_props, n0.depth AS n0_depth, n0.sort_order AS n0_sort_order FROM phpcr_nodes n0 WHERE n0.workspace_name = ? AND n0.type IN ('nt:unstructured', 'rep:root') AND EXTRACTVALUE(n0.props, 'count(//sv:property[@sv:name="phpcr:class"]/sv:value[text()="Symfony\\Cmf\\Bundle\\SeoBundle\\Model\\SeoMetadata"]) > 0')' with params ["default"]:

SQLSTATE[42S22]: Column not found: 1054 Unknown column 'n0.numerical_props' in 'field list'
```
Is that the right syntax for the query?

Ta

**Comment by liquorvicar** [comment]
In fact I just ran the UPDATE query as above and that failed too with the same error. Is this a bug in PHPCRSH?

**Comment by dbu** [comment]
i think the query you want could look like:

PHPCRSH > UPDATE [nt:unstructured]  WHERE
[phpcr:class]="Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata" SET
[phpcr:class]="Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetadata"

**Comment by liquorvicar** [comment]
No, sorry, no luck:
```
PHPCRSH > UPDATE [nt:unstructured] WHERE [phpcr:class]="Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata" SET [phpcr:class]="Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetadata"
[Doctrine\DBAL\Exception\InvalidFieldNameException] An exception occurred while executing 'SELECT n0.id AS n0_id, n0.path AS n0_path, n0.parent AS n0_parent, n0.local_name AS n0_local_name, n0.namespace AS n0_namespace, n0.workspace_name AS n0_workspace_name, n0.identifier AS n0_identifier, n0.type AS n0_type, n0.props AS n0_props, n0.numerical_props AS n0_numerical_props, n0.depth AS n0_depth, n0.sort_order AS n0_sort_order FROM phpcr_nodes n0 WHERE n0.workspace_name = ? AND n0.type IN ('nt:unstructured', 'rep:root') AND EXTRACTVALUE(n0.props, 'count(//sv:property[@sv:name="phpcr:class"]/sv:value[text()="Symfony\\Cmf\\Bundle\\SeoBundle\\Model\\SeoMetadata"]) > 0')' with params ["default"]:

SQLSTATE[42S22]: Column not found: 1054 Unknown column 'n0.numerical_props' in 'field list'
```

**Comment by dbu** [comment]
oh, this actually looks a lot like you upgraded jackalope-doctrine-dbal to version 1.2 without following https://github.com/jackalope/jackalope-doctrine-dbal/blob/master/UPGRADE.md - can you try that?

**Comment by liquorvicar** [comment]
Ok, looks like we have some action!
```
PHPCRSH > UPDATE [nt:unstructured] WHERE [phpcr:class]="Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata" SET [phpcr:class]="Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetadata"
[PHPCR\NamespaceException] Mapping for '[phpcr' is not defined
PHPCRSH > UPDATE [nt:unstructured] SET phpcr:class="Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetadata" WHERE phpcr:class="Symfony\Cmf\Bundle\SeoBundle\Model\SeoMetadata"
18 row(s) affected in 0.04s
```

And just checked `cache:clear` and it all works fine.

Many thanks for you help. Love you loads ;-)

**Comment by dbu** [comment]
\o/

leaving this open because we still should fix the form type to use the
right model by default

**Comment by liquorvicar** [comment]
I was just checking our app to see who in our team had upgraded jackalope-doctrine-dbal and hadn't run the upgrade sql and we have this in our composer.lock:
```
"name": "jackalope/jackalope-doctrine-dbal",
            "version": "1.1.5",
            "source": {
                "type": "git",
                "url": "https://github.com/jackalope/jackalope-doctrine-dbal.git",
                "reference": "90e3c706762fa2a1623b5d837d36cb1dad843a50"
            },
```
Isn't that a bit odd?

**Comment by dbu** [comment]
did you verify you really have the version you have in the lock file? i just checked the 1.1 branch and the specific 1.1.5 commit and the word `numerical_props` does not occur in the code.

**Comment by ElectricMaxxx** [comment]
Mit freundlichen Grüßen

Maximilian Berghoff


---------------------------------------
Maximilian Berghoff
Wiesenstraße 44
91617 Oberdachstetten 

Mail: Maximilian.berghoff@gmx.de
Mobile: +49 151 64825096

**Comment by liquorvicar** [comment]
Well, we're using composer in the standard way so I'd hope that the lock file does actually represent what we have installed!  `composer install --dry-run` doesn't list that package as due for an update on the system. I have manually double-checked the actual files on the staging server and there is no v1.2 listed in the CHANGELOG.md so I'm pretty sure we do actually have that version. Mind you, like you, when I grep the actual files for "numerical_props" nothing comes up. Should I raise an issue against phpcrsh?

**Comment by dbu** [comment]
that sounds really confusing. @dantleech @lsmith77 would you have an
idea why that error about numerical_props would pop up in this case?

**Comment by dantleech** [comment]
Just guessing: If you are using the PHPCRSH PHAR instead of embedding, then it is going to be using the Jackalope version with which it is compiled. In this case it looks like the version your app is running is older.

Try installing PHPCR Shell as a `dev` dependency instead and run it with `./app console doctrine:phpcr:shell` ?

**Comment by liquorvicar** [comment]
@dantleech Ah, yes, of course. That makes perfect sense. I downloaded the Beta 3 phar and from the current composer.json it looks like that is indeed compiled against 1.2. Thanks.

**Comment by ElectricMaxxx** [comment]
I missed the point by that stuff above, what would be the task of this issue in generall?

**Comment by dbu** [comment]
the thing is that we need to make sure that the form type is using the correct data_class and does not store the model class but the full phpcr class of seo metadata in the database when creating new content with metadata. 

the problem is that the form type is general: https://github.com/symfony-cmf/SeoBundle/blob/master/Form/Type/SeoMetadataType.php#L73 but when working with phpcr documents, we should use a different default. i don't know what the best solution is. you can either specify the class when using the form type, or have the document instantiate the metadata child to the correct object. but both of these are dangerous because they can be forgotten. ideally there would be something automatic like a guesser that sees what is going on.

**Comment by dbu** [comment]
fixed in #249
